crates/proving-service-client/build.rs

```diff
@@ -1,81 +1,61 @@
+    // Build nostd version
+    let nostd_path = format!("{GENERATED_OUT_DIR}/nostd");
+    build_tonic_from_descriptor(proving_service_descriptor, nostd_path.clone(), false)?;
 
-    fs::remove_dir_all(CRATE_PROTO_DIR).into_diagnostic()?;
-    fs::create_dir_all(CRATE_PROTO_DIR).into_diagnostic()?;
-    fs::copy(src_file, dest_file).into_diagnostic()?;
+    // Convert nostd version to use core/alloc instead of std
+    let nostd_file_path = format!("{nostd_path}/proving_service.rs");
+    convert_to_nostd(&nostd_file_path)?;
 
     Ok(())
 }
 
-fn compile_tonic_client_proto() -> miette::Result<()> {
-    let crate_root =
-        PathBuf::from(env::var("CARGO_MANIFEST_DIR").expect("CARGO_MANIFEST_DIR should be set"));
-    let dst_dir = crate_root.join("src").join("proving_service").join("generated");
-
-    // Remove `proving_service.rs` if it exists.
-    // We don't need to check the success of this operation because the file may not exist.
-    let _ = fs::remove_file(dst_dir.join("std").join("proving_service.rs"));
-    let _ = fs::remove_file(dst_dir.join("nostd").join("proving_service.rs"));
-
-    let out_dir = env::var("OUT_DIR").into_diagnostic()?;
-    let file_descriptor_path = PathBuf::from(out_dir).join("file_descriptor_set.bin");
-
-    let proto_dir: PathBuf = CRATE_PROTO_DIR.into();
-    let protos = &[proto_dir.join("proving_service.proto")];
-    let includes = &[proto_dir];
-
-    let file_descriptors = protox::compile(protos, includes)?;
-    fs::write(&file_descriptor_path, file_descriptors.encode_to_vec()).into_diagnostic()?;
+// HELPER FUNCTIONS
+// ================================================================================================
 
-    // Codegen for wasm transport and std transport
-    let nostd_path = dst_dir.join("nostd");
-    let std_path = dst_dir.join("std");
-    build_tonic_client(&file_descriptor_path, &std_path, protos, includes, false)?;
-    build_tonic_client(&file_descriptor_path, &nostd_path, protos, includes, true)?;
+/// Builds tonic code from a `FileDescriptorSet` with specified configuration
+fn build_tonic_from_descriptor(
+    descriptor: FileDescriptorSet,
+    out_dir: String,
+    build_transport: bool,
+) -> miette::Result<()> {
+    tonic_build::configure()
+        .out_dir(out_dir)
+        .build_server(false)
+        .build_transport(build_transport)
+        .compile_fds_with_config(prost_build::Config::new(), descriptor)
+        .into_diagnostic()
+}
 
-    // Replace `std` references with `core` and `alloc` in `proving_service.rs`.
-    // (Only for nostd version)
-    let nostd_file_path = nostd_path.join("proving_service.rs");
-    let file_content = fs::read_to_string(&nostd_file_path).into_diagnostic()?;
+/// Replaces std references with core and alloc for nostd compatibility
+fn convert_to_nostd(file_path: &str) -> miette::Result<()> {
+    let file_content = fs::read_to_string(file_path).into_diagnostic()?;
```

Comment:
I'm surprised there isn't support for this within `tonic` itself?

Reply:
Sadly I didn't find anything that that performs this directly from the tonic building. In the miden-client we are doing the same for the web client